### PR TITLE
[feat #607] Adding OFFSET sub command to the range queries.

### LINKED ISSUE #607: Add an offset for aggregation queries to allow for aligned results
Something I've been battling with lately, and it seems like there is no way I can set an offset when I'm aggregating with a timebucket. I'm using the python client.

To illustrate, lets assume I have ingested the data:
```python
In [1]: data
Out[1]:
[('test', 2, 1.585011688970841),
 ('test', 3, 1.5206503063325287),
 ('test', 4, 1.1504853132661446),
 ('test', 5, 1.6335210183200621),
 ('test', 6, 1.9591734441557134),
 ('test', 7, 1.9288271396549201),
 ('test', 8, 1.9699958302004554),
 ('test', 9, 1.205644987739304),
 ('test', 10, 1.7922129321935265),
 ('test', 11, 1.4875839601693452)]
```
This is data that starts at 2, and increase by 1 between values. 

Let's assume I want to get aggregated results from timestamp `5` until `9`, with steps of `3`. 
```python
In [2]: rts.range('test_2', from_time=5, to_time=9, aggregation_type="avg", bucket_size_msec=3)
Out[2]: [(3, 1.63352101832006), 
         (6, 1.95266547133703), 
         (9, 1.2056449877393)]
```

So, you can see that this doesn't start from `5`, but rather, it starts from `3`, and the value in the `3` bucket is the raw value for `5`, the next value is `6`... which makes sense, since it's `3` more than `3`. 

It would make sense to have the system allow for the aggregate to start from `5`, striding through data in steps of 3. 
```python
[(5, 1.84050720071023),
 (8, 1.58782040896988)]
```
In this case, the `5` timestamp will be the average of `5`, `6`, and `7` timestamps, 
and the `8` timestamp will be the average of the `8`, and `9` timestamps.

### PR DESCRIPTION
**Pull request by CaptainCuddleCube**
PR closes #607 

This PR adds an `OFFSET` subcommand (similar to the `COUNT` subcommand) to allow for a time series offset to be set. 

I decided against adding an extra option to the `AGGREGATE` command, because that would require downstream clients to change. By adding the `OFFSET` as it's own subcommand, nothing else changes (no tests fail), and setting the offset makes sense to be uncoupled from aggregation.

### CODE REVIEW

**Comment by CaptainCuddleCube** [comment]
Oooff, I forgot to format the code 🙈 . 

I will fix it now.

**Comment by CaptainCuddleCube** [comment]
I'm adding some tests now! Spotted a tiny corner case bug too!

File: src/module.c [hunk 1]
```diff
@@ -709,13 +740,14 @@ int ReplySeriesRange(RedisModuleCtx *ctx,
         timestamp_t init_ts = (rev == false)
                                   ? series->funcs->GetFirstTimestamp(iterator.currentChunk)
                                   : series->funcs->GetLastTimestamp(iterator.currentChunk);
-        last_agg_timestamp = init_ts - (init_ts % time_delta);
 
+        last_agg_timestamp = init_ts - (init_ts % time_delta) + time_offset;
```
Comment (danni-m) [inline src/module.c:744]: Can you explain the logic behind this? i would expect to see the following:
```
        last_agg_timestamp = init_ts - (init_ts % time_delta) + time_offset;
```

File: docs/commands.md [hunk 2]
```diff
@@ -229,15 +229,17 @@ Note: Whenever filters need to be provided, a minimum of one `l=v` filter must b
...
+TS.REVRANGE key fromTimestamp toTimestamp [COUNT count] [OFFSET offset] [AGGREGATION aggregationType timeBucket]
 ```
 
 - key - Key name for timeseries
 - fromTimestamp - Start timestamp for the range query. `-` can be used to express the minimum possible timestamp (0).
 - toTimestamp - End timestamp for range query, `+` can be used to express the maximum possible timestamp.
 
 Optional args:
+* count - Maximum number of returned results per time-series.
+* offset - Offset the timestamps being returned; allows for the starting value to be set.
```
Comment (danni-m) [inline docs/commands.md:242]: We also want to mention this is millisecond based offset to be aligned with other input/output related to timestamps

File: tests/flow/test_ts_mrange.py [hunk 3]
```diff
@@ -83,6 +83,15 @@ def build_expected(val, time_bucket):
                                           'FILTER', 'generation=x')
         assert 18 == len(actual_result[0][2])  # just checking that agg count before count works
 
+        actual_result = r.execute_command('TS.mrange', start_ts+2, -1, 'COUNT', 1, 'OFFSET', 2,
+                                          'FILTER', 'generation=x')
+        assert start_ts + 4==actual_result[0][2][0][0] 
+
+        # Note that we need to have the offset synchronized to the aggregation to allow for consistent offset values.
+        actual_result = r.execute_command('TS.mrange', start_ts+2, -1, 'COUNT', 1, 'OFFSET', (start_ts + 5) % 3,
```
Comment (danni-m) [inline tests/flow/test_ts_mrange.py:91]: I think we want to have something more rigorous here to test the full response and verify all the buckets are working as expected.

**Review by danni-m — COMMENTED** [review]
@CaptainCuddleCube Thanks for your contribution!
Sorry for the late response, we appreciate the work you did here.

First, I would like to understand better the use-case for this (we had a few requests for this already, but its always good to hear all voices).

The most of it looks OK to me, there are several issues I would like to raise:
1. For aggregated queries, are we adding the timestamp post bucketing or prior? i see that you chose to add it after bucketing and I would like to understand the reasoning.
2. We need to also test negative offsets.
3. Testing and validating a full response.

Thanks @CaptainCuddleCube !

File: src/module.c (hunk 1 above)
Comment (CaptainCuddleCube) [reply src/module.c:744]: I'm not sure I follow?

Comment (danni-m) [reply src/module.c:744]: sorry pasted the wrong line, i meant to paste this:
```
  last_agg_timestamp = init_ts - ((init_ts + time_offset) % time_delta);
  ```

**Comment by CaptainCuddleCube** [comment]
> @CaptainCuddleCube Thanks for your contribution!
> Sorry for the late response, we appreciate the work you did here.
> 
> First, I would like to understand better the use-case for this (we had a few requests for this already, but its always good to hear all voices).
> 
> The most of it looks OK to me, there are several issues I would like to raise:
> 
> 1. For aggregated queries, are we adding the timestamp post bucketing or prior? i see that you chose to add it after bucketing and I would like to understand the reasoning.
> 2. We need to also test negative offsets.
> 3. Testing and validating a full response.
> 
> Thanks @CaptainCuddleCube !

Hey Danny! 
Thank you for the PR. 

So, specifically, the use case I'm after is to align the aggregation to the requested start time. The reason why this is important, is we are using RTS as a write-through cache so that we don't have to hit our TimescaleDB for numeric time series information. With TimescaleDB, we can "offset" the time, or set what the starting timestamp should be; so that the aggregation gaps will be relative to that time. The ideal case for me is to have our API read aggregated responses from RTS when it can, and only when there is cache miss, read from TimescaleDB; so if we can get the two outputs looking the same, we have massive performance gains.

I opted for setting a generic offset - since this allows for the most flexility. 

1. So, for this, it's totally up to you. I wasn't really sure what made the most sense; and I was spending more time spinning wheels trying to wrap my head around it. I should have made this clear in my PR comment. 
2. Absolutely; that slipped through the cracks!
3. What do you mean by "full response"?

File: src/module.c (hunk 1 above)
Comment (CaptainCuddleCube) [reply src/module.c:744]: So, I tried that, but it I couldn't get sane aggregations out, and I was struggling to make it work for `RANGE` and `REVRANGE`. 

However; the last time I worked on a C project was about 6 years ago; so I was a little rusty. 

I'm happy to change anything, and double check things on my side!

**Comment by CaptainCuddleCube** [comment]
@danni-m 

I need to rebase this, since the main branch has drifted and there's been some changes. I should be able to get this out in the next couple days.

**Comment by danni-m** [comment]
Hi @CaptainCuddleCube, we've added a new feature to support your use-case using the `ALIGN` parameter. please take a look at the docs: https://oss.redislabs.com/redistimeseries/master/commands/#tsmrangetsmrevrange

**Comment by CaptainCuddleCube** [comment]
> Hi @CaptainCuddleCube, we've added a new feature to support your use-case using the `ALIGN` parameter. please take a look at the docs: https://oss.redislabs.com/redistimeseries/master/commands/#tsmrangetsmrevrange

Amazing! thank you! I think we can close this ticket!
